added renameLast and verbose options

azimonti opened the pull request:
<!--
Thank you for creating a pull request to contribute to Hexo!
-->

## check list

- [x] Add test cases for the changes.
- [ ] Passed the CI test.

## Description

Adding the option to have the last page to be latest which is convenient in case of reverse pagination.
Adding a verbose option to display the generated routes

## Additional information

azimonti committed:
added overwriteLatest and verbose options

uiolee commented on lib/pagination.js:
```diff
@@ -78,6 +80,16 @@ function pagination(base, posts, options = {}) {
 
+  if ((overwriteLatest && result.length > 1) || (overwriteLatest && explicitPaging && result.length === 1)) {
+    const lastPage = result[result.length - 1];
+    lastPage.path = lastPage.path.replace(/\/page\/\d+\/?$/, '/latest/');
```
https://github.com/hexojs/hexo-pagination?tab=readme-ov-file#usage

The url format may be diverse, and this matching method is not appropriate.

uiolee replied:
https://github.com/hexojs/hexo-pagination/blob/0804e679f9bc7672bfa2decd8f77f92ed5c95de8/README.md?plain=1#L44-L54

Each page's data contains `current_url` `prev_link` `next_link`. Only modifying the `path` will cause wrong reference

uiolee commented on README.md:
```diff
@@ -16,13 +16,17 @@ $ npm install hexo-pagination --save
+| `layout`         | Layout. This value can be a string or an array.         | `['archive', 'index']` |
+| `data`           | Extra data                                              | `{}`                   |
+| `explicitPaging` | Number the first page. e.g. `page/1/index.html`         | `false`                |
+| `overwriteLatest`| Set the latest page name. e.g. `page/latest/index.html` | `false`                |
```
I don't know if you have done any relevant testing. By default, articles are sorted in reverse order by date, so the first page is "latest". 

However, different users may have different settings, so I would suggest that two options should be provided, freely choosing to overwrite the first page or the last page. 

For this PR, I would recommend that call it "lastpage" instead of "latest"

azimonti replied:
the objective is to have the article sorted in normal order, so page 1 has the oldest post and never changing.
reverse order is regenerating all pages because the order is changing.
my objective is to minimize the number of pages which are generated. With normal order it generates only 1 page each day, and not tenth of pages.

I start from the last page and navigate backward. 

To achieve this I need to link the last page on my home, or I need to change the page link every time there is a new page.

uiolee commented on test/index.js:
```diff
@@ -152,4 +152,29 @@ describe('pagination', () => {
+    }
+    result[result.length - 1].path.should.eql('/latest/');
+  });
+  it('overwriteLatest 2', () => {
```
It is recommended to summarize it, instead of naming it 123

azimonti committed:
simplified the logic of renameLast

azimonti commented:
thank you for the many comments and suggestions. 

I made few changes:
* simplified the logic directly within formatURL to cover all the options
* changed the name from `latest` to `last`, but added option for user to change it as they prefer
* added more tests (including localization and different format) and consolidate the tests in one test function

uiolee commented on lib/pagination.js:
```diff
@@ -26,7 +29,9 @@ function pagination(base, posts, options = {}) {
     let url;
-    if (!explicitPaging) {
+    if (renameLast && explicitPaging && i === total) {
+      url = base + format(_format, i).replace(i, localizedLast);
```
`.replace(i, localizedLast)`

this still a potential problem, it may replace other content in the url.

```suggestion
      url = base + _format.replace('%d', localizedLast);
```

uiolee commented on test/index.js:
```diff
@@ -152,4 +152,37 @@ describe('pagination', () => {
+    }
+    result[result.length - 1].path.should.eql('/page/last/');
+    result[result.length - 2].data.next_link.should.eql('/page/last/');
+    result = pagination('/', posts, {
```
Please avoid putting multiple tests in one `it`. They should be separated, not merged

azimonti replied:
last time you asked me 
> It is recommended to summarize it, instead of naming it 123

so I summarized in one test. let me split again.

azimonti replied:
I changed the tests

```
    pagination renameLast
      ✔ should generate /page/{num}/ paths with explicitPaging and renameLast for multiple posts
      ✔ should work with localizedLast option with explicitPaging and renameLast
      ✔ should produce /page/last/ for a single post with explicitPaging and renameLast
      ✔ should produce root path for a single post without explicitPaging even if renameLast is set
```

uiolee commented on README.md:
```diff
@@ -16,13 +16,18 @@ $ npm install hexo-pagination --save
+| `localizedLast`  | Localize the last page name. e.g. `page/最後/index.html` | `last`                 |
+| `verbose`        | Display the generated urls                               | `false`                |
+
+If there is a single page `overwriteLatest` requires `explicitPaging=true`.
```
According to your code, `renameLast` always need `explicitPaging`

https://github.com/hexojs/hexo-pagination/blob/ae908ce2d17273ca76fff13afda8ef753247bfa8/lib/pagination.js#L32

azimonti replied:
yes. 
in the case `explicitPaging=false` and `localizedLast=true`, if there is a single page like:

`/category/tag/mytag/`. In this case, the user is expecting:
- `/category/tag/mytag/` (`explicitPaging=false` -> not touch if there is a single page)
-  `/category/tag/mytag/last` (`localizedLast=true` -> the last page need to be changed)
I cannot see which is the user expected behaviour

conversely`explicitPaging=true` and `localizedLast=true`, what the use expect is clear.

azimonti committed:
Update lib/pagination.js

Co-authored-by: Uiolee <22849383+uiolee@users.noreply.github.com>
Signed-off-by: Marco Azimonti <git@azimonti.com>

azimonti committed:
Separate multiple pagination tests into individual cases for clarity and maintainability

azimonti reviewed (commented):
merged the review
split test cases

azimonti reviewed (commented):
reviewed
